[Core] Cache ProjectFile.Include to improve project save performance

mrward opened the pull request:
With an SDK style project containing 1500 C# files, adding a new
C# class to the project would result in the saving taking about 20
seconds. The majority of this time was spent in the Project's
SaveProjectItems as it tried to determine if an MSBuild remove item
should be added for the new file. Caching the ProjectFile.Include
takes the save time down to around 300 ms.

Fixes VSTS #947103 - Class creation is very slow in projects with
hundreds of classes

mrward committed:
[Core] Cache ProjectFile.Include to improve project save performance

With an SDK style project containing 1500 C# files, adding a new
C# class to the project would result in the saving taking about 20
seconds. The majority of this time was spent in the Project's
SaveProjectItems as it tried to determine if an MSBuild remove item
should be added for the new file. Caching the ProjectFile.Include
takes the save time down to around 300 ms.

Fixes VSTS #947103 - Class creation is very slow in projects with
hundreds of classes

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ProjectFile.cs:
```diff
@@ -485,12 +495,31 @@ internal bool ResolveParent ()
+
 				VirtualPathChanged?.Invoke (this, new ProjectFileVirtualPathChangedEventArgs (this, FilePath.Null, ProjectVirtualPath));
+			} else {
+				if (project != null)
```
We should unsubscribe this event handler always, not just on this code-path, to prevent leaking the project when `Project` is set multiple times.

Therzok replied:
Also, since ProjectItem does not compare the project instances being set, even setting the same value twice will subscribe `OnProjectModified` _twice_ to the project.

mrward replied:
Good catch 👍

Therzok commented on main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/DotNetCoreProjectTests.cs:
```diff
@@ -696,6 +696,44 @@ public async Task LoadProject_UnknownNuGetSDKPackage_SDKResolutionErrorsReported
+				// This was taking 20 seconds before.
+				// Takes about 300ms with ProjectFile.Include caching. Here we use 2 seconds
+				// in case the build server is slow.
+				Assert.That (timer.ElapsedMilliseconds, Is.LessThan (2000));
```
Should we add a new TimerCounter for `SaveAsync` and use that information here?

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/ProjectFile.cs:
```diff
@@ -485,15 +495,31 @@ internal bool ResolveParent ()
 		{
 			base.OnProjectSet ();
+			if (project != null)
+				project.Modified -= OnProjectModified;
```
```csharp
if (project != null) {
    project.Modified -= OnProjectModified;
    project = null;
}
```

mrward replied:
😄

slluis commented:
An idea to simplify the code and avoid too many handlers for the `Project.Modified` event: keep a reference of the `Project.ItemDirectory` value that was used to create `cachedInclude` in ProjectFile, and return `cachedInclude` only if  `Project.ItemDirectory` has not changed.

mrward commented:
I originally tried keeping the original value of the Project.ItemDirectory (both as a FilePath and as a string) and comparing the new value with the old value, to see if it had changed, but that did not seem to fix the performance problem enough. Doing that dropped the save timing down to about 7 or 8 seconds. So I switched to using the Project.Modified event for better performance.

slluis commented:
Wow, that means this property is invoked a huge amount of times. It is a perfect candidate for a performance test.

slluis commented:
@monojenkins backport release-8.2
